### PR TITLE
fix(ext/ffi): trampoline for fast calls

### PR DESCRIPTION
**Pull request by littledivy**
Pointed out by @aapoalas #15125 introduced a bug in FFI calls. The first parameter of a fast function must be a V8Value receiver. This patch implements a fix by patching symbol calls to ignore the receiver.

### CODE REVIEW

File: ext/ffi/jit_trampoline.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      can_callback: false,
+    };
+    // undefined symbol _func
+    assert!(gen_trampoline(Box::new(sym)).is_err());
+  }
+
+  #[test]
+  fn test_gen_trampoline() {
+    assert_eq!(
+      codegen(vec![], NativeType::Void),
```
Comment (aapoalas) [inline ext/ffi/jit_trampoline.rs:123]: Nice! I like seeing tests with this.

File: ext/ffi/jit_trampoline.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+macro_rules! cstr {
+  ($st:expr) => {
+    &CString::new($st).unwrap()
+  };
+}
+
+fn native_to_c(ty: &NativeType) -> &'static str {
+  match ty {
+    NativeType::U8 | NativeType::U16 | NativeType::U32 => "unsigned int",
+    NativeType::I8 | NativeType::I16 | NativeType::I32 => "int",
```
Comment (aapoalas) [inline ext/ffi/jit_trampoline.rs:24]: I think we should extend support for 64bit `typeof arg === "number"` parameters. The Fast API supports those, and we support those in our normal FFI calls; we'll always return BigInts but we do allow users to pass in a number as a u64 / i64 / usize / isize. They're forced into only up-to 53bit values, but that's what they have in the JS numbers originally anyway so there's no risk of losing data or anything like that.

And if they need the full 64bits to represent the number, they'll use a BigInt which means that V8 will choose to not call the Fast API function.

File: ext/ffi/lib.rs [hunk 3]
```diff
@@ -702,19 +705,24 @@ fn make_sync_fn<'s>(
...
+      jit_trampoline::gen_trampoline(sym.clone()).expect("gen_trampoline");
     fast_ffi_templ = Some(FfiFastCallTemplate {
       args: args.into_boxed_slice(),
-      ret: (&fast_api::Type::from(&sym.result_type)).into(),
-      symbol_ptr: sym.ptr.as_ptr() as *const c_void,
+      ret: (&ret).into(),
+      symbol_ptr: symbol_trampoline.addr,
     });
+    // Leak.
+    Box::leak(symbol_trampoline);
```
Comment (aapoalas) [inline ext/ffi/lib.rs:725]: Should this be bound to a finalizer?

**Comment by aapoalas** [comment]
Oh yeah, regarding the u8 / u16 / i8 / i16 support, @Phosra had a good note about return values:
https://github.com/denoland/deno/pull/15122#discussion_r917323197

File: ext/ffi/jit_trampoline.rs (hunk 2 above)
Comment (littledivy) [reply ext/ffi/jit_trampoline.rs:24]: Sure, let's do this in a follow up. I want to keep this patch as a 'fix'.

**Comment by aapoalas** [comment]
> If the first parameter is now the receiver, does this mean that C functions with empty parameters would now supported?

0-ary C functions were already supported. eg. Here're a few lines from benchmarks I ran on a PR of mine which changes nothing related to normal FFI calls:
```
benchmark                              time (avg)             (min … max)       p75       p99      p995
------------------------------------------------------------------------- -----------------------------
nop()                                3.56 ns/iter    (3.45 ns … 19.23 ns)    3.5 ns   4.37 ns   4.75 ns
add_u32()                            7.91 ns/iter    (7.17 ns … 18.92 ns)   7.64 ns  13.32 ns  15.29 ns
nop_u8()                             7.88 ns/iter     (7.4 ns … 34.05 ns)    7.4 ns  12.53 ns  16.59 ns
```

This just changes how those calls are called: Currently they're given an extra parameter that they do not expect but luckily also do not care about.

File: ext/ffi/jit_trampoline.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  }
+  c += ");\n\n";
+
+  // void* recv, <param_type> p0, <param_type> p1, ...);
+  c += ret;
+  c += " func_trampoline(";
+  c += "void* recv";
+  for (i, ty) in sym.parameter_types.iter().enumerate() {
+    c += ", ";
+    c += native_to_c(ty);
```
Comment (aapoalas) [inline ext/ffi/jit_trampoline.rs:60]: It'd be more accurate here to use a different type mapping for the trampoline, since V8 will never actually call with u8s, u16s etc. All of those parameter types should become 32 bit numbers, and corresponding casting & masking should be applied to convert the parameter value to the actual FFI call parameter value.

File: test_ffi/tests/test.js [hunk 5]
```diff
@@ -388,13 +400,13 @@ dylib.symbols.call_stored_function_2(20);
...
-    dylib.symbols.call_stored_function_2(10);
+    symbols.call_stored_function_2(10);
   },
 );
-dylib.symbols.store_function(ptr(nestedCallback));
+symbols.store_function(ptr(nestedCallback));
 
-dylib.symbols.store_function(null);
-dylib.symbols.store_function_2(null);
+symbols.store_function(null);
```
Comment (aapoalas) [inline test_ffi/tests/test.js:408]: Why the `dylib.` removals? Does this allow the other calls to be optimised as well? If not, then maybe revert this change as unnecessary.

File: ext/ffi/jit_trampoline.rs (hunk 4 above)
Comment (aapoalas) [reply ext/ffi/jit_trampoline.rs:60]: V8's Fast API supported types can be referenced here: https://v8docs.nodesource.com/node-18.2/da/d05/classv8_1_1_c_type_info.html (This is just a very rough API doc, it doesn't really explain much. Not all of the types are also supported in all positions.)

As evident, only 32bit and 64bit integers are mentioned.

Beyond that, u8 / u32 casting and masking probably needs its own reference :)

File: ext/ffi/lib.rs [hunk 6]
```diff
@@ -72,6 +77,7 @@ struct Symbol {
   ptr: libffi::middle::CodePtr,
   parameter_types: Vec<NativeType>,
   result_type: NativeType,
+  #[allow(dead_code)]
   can_callback: bool,
```
Comment (ry) [inline ext/ffi/lib.rs:81]: What's happening here?

**Review by ry — COMMENTED** [review]
It would be very nice to have ext/ffi/README updated with an summary of how this works.

File: test_ffi/tests/test.js (hunk 5 above)
Comment (ry) [reply test_ffi/tests/test.js:408]: If unnecessary, I agree. This is a complex patch and should be kept as minimal as possible.

File: ext/ffi/lib.rs (hunk 6 above)
Comment (littledivy) [reply ext/ffi/lib.rs:81]: yeah, its dead code in Windows for now.

File: test_ffi/tests/test.js (hunk 5 above)
Comment (littledivy) [reply test_ffi/tests/test.js]: Reverted

File: ext/ffi/lib.rs (hunk 6 above)
Comment (ry) [reply ext/ffi/lib.rs:81]: Please add a comment

**Review by ry — COMMENTED** [review]
Please add an explanation of the process to ext/ffi/README

**Review by ry — APPROVED** [review]
LGTM
